cni/pkg/cmd/root.go

```diff
@@ -105,6 +105,8 @@ var rootCmd = &cobra.Command{
 					DNSCapture:        cfg.InstallConfig.AmbientDNSCapture,
 					EnableIPv6:        cfg.InstallConfig.AmbientIPv6,
 					TPROXYRedirection: cfg.InstallConfig.AmbientTPROXYRedirection,
+					Reconcile:         cfg.InstallConfig.AmbientReconcile,
```

Comment:
Let's collapse these in to a single ambient option - 

`ReconcilePodRulesOnStartup`

so that people can turn off that singular option if they want current behavior.

Reply:
Do you suggest to collapse Reconcile and RepairReconcile? Or also ForceApply?

I propose keeping ForceApply separate. It’s a safety flag unrelated to reconciliation, intended to disable the new idempotent behavior of the iptables library. The flag will be removed from both CNI and non-CNI iptables at a later date.
Under normal circumstances, there’s no reason to set this flag to true unless a user needs a workaround due to a bug in the idempotency logic.
When ForceApply is set to true, the iptables library will apply the rules even if identical rules are already present, whereas, with the new default behavior, it would normally skip reapplying them.

Reply:
> Under normal circumstances, there’s no reason to set this flag to true unless a user needs a workaround due to a bug in the idempotency logic.

In that case, since this entire feature is net-new and already behind a feature flag, I would not expose this flag at all here.

Reply:
Added ReconcilePodRulesOnStartup.

Keep in mind that this PR also introduces the same idempotent handling used in tools/istio-iptables, which prevents the same set of rules from being applied twice. This handling covers all use cases, including the scenario where ReconcilePodRulesOnStartup=false.
However, users will not see any functional change even for ReconcilePodRulesOnStartup=false, as the approach for host rules and in-pod rules on the master node make sure they are applied in a clean state, automatically resulting in a standard first-time installation.
This holds true unless an unlikely bug causes the idempotency logic to incorrectly interpret the existing state as the desired one leading to a skip of the apply step. In such cases, ForceApply serves as a safety measure to bypass that logic.

If ReconcilePodRulesOnStartup is set to true, then preventing the same set of rules from being applied twice (if they are already compatible) becomes mandatory, hence ReconcilePodRulesOnStartup=true will always make sure that ForceApply is set to false, regardless of what the user specified.

Reply:
> Added ReconcilePodRulesOnStartup.
> 
> Keep in mind that this PR also introduces the same idempotent handling used in tools/istio-iptables, which prevents the same set of rules from being applied twice. This handling covers all use cases, including the scenario where ReconcilePodRulesOnStartup=false. However, users will not see any functional change even for ReconcilePodRulesOnStartup=false, as the approach for host rules and in-pod rules on the master node make sure they are applied in a clean state, automatically resulting in a standard first-time installation. 

That's fine/desirable. Existing ambient behavior is that we never try to write iptables rules in the pod after we have already done so once and annotated the pod. There are pathological situations where user intervention can create this problem (people manually delete the annotation the control plane uses to track enrollment state, we try to renroll, and fail because the rules are already there). That's pretty rare/pathological but preventing it is a nice net-new feature, and isn't something we need to worry about feature-flagging.

> This holds true unless an unlikely bug causes the idempotency logic to incorrectly interpret the existing state as the desired one leading to a skip of the apply step. In such cases, ForceApply serves as a safety measure to bypass that logic.

Yes - we will likely never need to set this in the `istio-cni` agent tho, since restarting the pod, or unenrolling and re-enrolling is always an option, and probably easier for people to do as a workaround if we do encounter such a bug - with ambient there are already plenty of workarounds for this, so we will never need `ForceApply=true`.

> If ReconcilePodRulesOnStartup is set to true, then preventing the same set of rules from being applied twice (if they are already compatible) becomes mandatory, hence ReconcilePodRulesOnStartup=true will always make sure that ForceApply is set to false, regardless of what the user specified.

Yes - that's my point - `ForceApply` will be `false` always in both cases, and for any breakage/bug it could conceivably work around, we already have several other short-term workarounds for.

So I would prefer to hardcode it to false and not expose it at all here.

Reply:
Or - at _least_ I would prefer not to put it in the chart values/configmap - the latter now supports setting arbitrary env, so if `FORCE_APPLY` is ever desired as a break-glass for a short-term bugfix, it's already possible to set without making it a user-facing Helm toggle.

Reply:
> That's fine/desirable. Existing ambient behavior is that we never try to write iptables rules in the pod after we have already done so once and annotated the pod. There are pathological situations where user intervention can create this problem (people manually delete the annotation the control plane uses to track enrollment state, we try to renroll, and fail because the rules are already there). That's pretty rare/pathological but preventing it is a nice net-new feature, and isn't something we need to worry about feature-flagging.

Aaah I did not expect this state to occurr but I did not anticipate users to actively mess with the internal annotations :P
Good to know that the new behavior would mitigate that.

I’ve removed the ForceApply feature from user-facing charts/configmaps and from AmbientArgs as well, as I doubt users will need it now.
ForceApply would only become useful if multiple bugs were present in the logic at once and I believe the new test coverage already prevents this possibility.

I’ll still keep the ForceApply flag in the CNI iptables library to maintain feature parity with tools/istio-iptables. I plan to remove the flag for both cni/istio-iptables and tools/istio-iptables around Istio 1.26, once it has been in place for about two releases.